src/libraries/System.Private.CoreLib/src/System/IO/FileStream.Unix.cs

```diff
@@ -657,6 +657,14 @@ private unsafe void WriteNative(ReadOnlySpan<byte> source)
                     offset += bytesWritten;
                     count -= bytesWritten;
                 }
+
+                // The POSIX 'write' method does not return ENOSPC when the number of bytes written is smaller than the
+                // number of bytes requested to write, but when that happens, it could mean there was not enough disk space.
+                // But'write' will not always return bytesWritten < count when there is not enough disk space, so there is no
+                // guarantee that the data can be commited to disk when the stream is disposed.
+                // To ensure we detect the insufficient disk error during disposal (not during finalization), we should call
+                // 'fsync' after calling 'write'.
+                FlushOSBuffer();
```

Comment:
This looks very expensive. What is this going to do to performance?

Reply:
Flush on every write is also a behavior change, effectively making the stream always unbuffered and no-oping OS buffers.